pkg/controller/servicediscovery/privatednsnamespace/setup.go

```diff
@@ -0,0 +1,55 @@
+package privatednsnamespace
+
+import (
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	svcsdk "github.com/aws/aws-sdk-go/service/servicediscovery"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/servicediscovery/v1alpha1"
+)
+
+// SetupPrivateDNSNamespace adds a controller that reconciles PrivateDNSNamespaces.
+func SetupPrivateDNSNamespace(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
+	name := managed.ControllerName(svcapitypes.PrivateDNSNamespaceGroupKind)
+	opts := []option{
+		func(e *external) {
+			e.observe = e.observeImpl
```

Comment:
It would be best to create a `hook` type that has `observe`, `create`, `update` and `delete`.

See https://github.com/crossplane/provider-aws/blob/master/pkg/controller/secretsmanager/secret/setup.go as example

Reply:
Thanks for the reference! I've updated the code to use a hooks struct.